### PR TITLE
[HYDRATOR-1239] Added support for Azure, Excel,FTP, File, S3 batch source plugins

### PR DESCRIPTION
**Pull request by romy-khetan**
Not able to test the changes for Azure, S3 and FTP changes due to the unavailability of clusters, unit tests and ITN.

### CODE REVIEW

**Comment by wolf31o2** [comment]
Azure and S3 are services. Your team has credentials for S3. Contact devops via email and we'll provide Azure credentials to the necessary capabilities. As for FTP, you can use Coopr to spin up a Docker cluster with an FTP service, using the [nikkor/pure-ftpd-with-user](https://hub.docker.com/r/nikkor/pure-ftpd-with-user/) image. You can connect to this container using the default username and password documented.

**Comment by romy-khetan** [comment]
@russorat Please suggest if we should go ahead with the suggestions provided by Chris.

File: core-plugins/docs/Excel-batchsource.md [hunk 1]
```diff
@@ -32,6 +32,12 @@ in **filePath** input. (Macro-enabled)
 **memoryTableName:** KeyValue table name to keep the track of processed files. This can be
 a new table or existing one. (Macro-enabled)
 
+**memoryDatasetExplorer:** Name of the Explorer table corresponding to 'memoryTableName' dataset. If not mentioned,
```
Comment (russorat) [inline core-plugins/docs/Excel-batchsource.md:35]: let's change this (and all other similar descriptions) to: "Name of the Hive table for the 'memoryTableName' dataset. By default, it is the same name as the dataset. (Macro-enabled)"

File: core-plugins/docs/Excel-batchsource.md [hunk 2]
```diff
@@ -32,6 +32,12 @@ in **filePath** input. (Macro-enabled)
...
+name will be same as the 'memoryTableName' dataset. (Macro-enabled)
+
+**memoryDatasetExplorerDatabase:** Existing Database name for the 'memoryDatasetExplorer' dataset.
```
Comment (russorat) [inline core-plugins/docs/Excel-batchsource.md:38]: Existing Hive database name for the 'memoryDatasetExplorer' dataset. By default, the table will be stored in the current namespace's database. (Macro-enabled)

File: core-plugins/docs/S3-batchsource.md [hunk 3]
```diff
@@ -22,7 +22,6 @@ Properties
 **accessKey:** Access Key of the Amazon S3 instance to connect to. (Macro-enabled)
 
 **path:** Path to file(s) to be read. If a directory is specified,
-terminate the path name with a '/'. The path uses filename expansion (globbing) to read files. (Macro-enabled)
```
Comment (russorat) [inline core-plugins/docs/S3-batchsource.md:25]: why was this line deleted?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/ExcelInputReader.java [hunk 4]
```diff
@@ -312,12 +313,21 @@ private void createDatasets(@Nullable PipelineConfigurer pipelineConfigurer, @Nu
         Map<String, String> properties = new HashMap<>();
         properties.put(Properties.Table.PROPERTY_SCHEMA, errorRecordSchema.toString());
         properties.put(Properties.Table.PROPERTY_SCHEMA_ROW_FIELD, KEY);
-        DatasetProperties datasetProperties = DatasetProperties.builder().addAll(properties).build();
+        ExploreProperties.Builder builder = ExploreProperties.builder();
+        if (!Strings.isNullOrEmpty(excelInputreaderConfig.errorDatasetExplorer)) {
```
Comment (russorat) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/ExcelInputReader.java:317]: wouldn't it be easier to set these in the default constructor so you could simplify this logic? you could set default values for everything so you could just have `builder.setExploreTableName(excelInputreaderConfig.errorDatasetExplorer);`

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/ExcelInputReader.java [hunk 5]
```diff
@@ -328,11 +338,23 @@ private void createDatasets(@Nullable PipelineConfigurer pipelineConfigurer, @Nu
 
       if (!excelInputreaderConfig.containsMacro("memoryTableName") &&
         !Strings.isNullOrEmpty(excelInputreaderConfig.memoryTableName)) {
+
+        ExploreProperties.Builder builder = ExploreProperties.builder();
+        if (!Strings.isNullOrEmpty(excelInputreaderConfig.memoryDatasetExplorer)) {
```
Comment (russorat) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/ExcelInputReader.java:343]: this logic is repeated multiple times. use defaults in the config constructors to remove this

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FileBatchSource.java [hunk 6]
```diff
@@ -134,7 +150,17 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
   public void prepareRun(BatchSourceContext context) throws Exception {
     // Need to create dataset now if macro was provided at configure time
     if (config.timeTable != null && !context.datasetExists(config.timeTable)) {
-      context.createDataset(config.timeTable, KeyValueTable.class.getName(), DatasetProperties.EMPTY);
+      ExploreProperties.Builder builder = ExploreProperties.builder();
+      if (!Strings.isNullOrEmpty(config.timeTableExplorer)) {
+        builder.setExploreTableName(config.timeTableExplorer);
+      } else {
```
Comment (russorat) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FileBatchSource.java:156]: duplicate code. figure out a way to have this code once and call it multiple times.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/ExcelInputReader.java (hunk 4 above)
Comment (romy-khetan) [reply core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/ExcelInputReader.java:317]: Value for 'errorDatasetExplorer' is based on value of 'errorDatasetName' and there should be no default value for 'errorDatasetName', that's the reason 'errorDatasetExplorer' could not be set in the default constructor.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/ExcelInputReader.java (hunk 5 above)
Comment (romy-khetan) [reply core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/ExcelInputReader.java:343]: Value for 'memoryDatasetExplorer' is based on value of 'memoryTableName' and there should be no default value for 'memoryTableName', that's the reason 'memoryDatasetExplorer' could not be set in the constructor.

**Comment by russorat** [comment]
closing due to age
